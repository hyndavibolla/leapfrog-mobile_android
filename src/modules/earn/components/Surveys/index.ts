export { default as Surveys } from './Surveys';
