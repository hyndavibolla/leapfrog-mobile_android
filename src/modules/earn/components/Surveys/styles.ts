import { css } from '@emotion/native';

import { COLOR } from '_constants';

export const styles = {
  sectionMain: css`
    padding-horizontal: 16px;
    padding-bottom: 24px;
  `,
  surveySectionHeader: css`
    padding-bottom: 16px;
    justify-content: space-between;
    flex-direction: row;
    align-items: center;
  `,
  link: css`
    color: ${COLOR.PRIMARY_BLUE};
  `
};
