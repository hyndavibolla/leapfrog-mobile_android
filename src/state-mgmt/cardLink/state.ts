import { ICardLinkOffer, ILinkedCard, ILocation } from '_models/cardLink';

export interface ICardLinkState {
  userId: string;
  offers: ICardLinkOffer[];
  linkedCardsList: ILinkedCard[];
  lastSearchedLocation: ILocation;
  isLocalOfferFailed: boolean;
  hasCalledLocalOffer: boolean; // to check if at least one time has been called getLocalOffers
  routeToActivateLocalOffer: string; // to activate the local offer after activated a new credit card
}

export const initialState: ICardLinkState = {
  userId: null,
  offers: [],
  linkedCardsList: [],
  lastSearchedLocation: null,
  isLocalOfferFailed: false,
  hasCalledLocalOffer: false,
  routeToActivateLocalOffer: ''
};
