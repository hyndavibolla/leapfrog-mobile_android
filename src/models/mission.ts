import { HTMLString, DateLike } from './general';
import { ProgramType } from './offer';

export interface IMission {
  uuid: string; // generated on the client (api service) by combining offerCode with offerId
  name: string;
  description: string;
  image?: string;
  termsAndConditions: HTMLString;
  brandName: string;
  brandLogo?: string;
  brandDescription: string;
  brandCategories: string[];
  brandRequestorId: string;
  offerId: string;
  offerCode: string;
  rank: number;
  validFrom: DateLike;
  validUntil: DateLike;
  pointsAwarded: IPointsAwarded;
  callToActionUrl: string;
  provider: Provider;
  pubRef: string;
  rewardOfferCode: string;
}

export enum Provider {
  BUTTON = 'button',
  SURVEY = 'AdBloom',
  CARDLINK = 'mastercard'
}

export enum RedemptionType {
  POINT_PER_DOLLAR = 'PointsPerDollar',
  FIXED_POINTS = 'FixedPoints',
  PERCENT_OFF = 'PercentOff'
}

export interface IPointsAwarded {
  rewardType: RedemptionType;
  rewardValue: number;
  conditions: IAwardCondition[];
  alternateRewardValue?: number;
  programType?: ProgramType;
  programSubType?: string;
}

export interface IAwardCondition {
  category?: string;
  rewardValue: number;
  rewardType: RedemptionType;
  rewardOfferCode?: string;
  programType?: ProgramType;
  programSubType?: string;
  alternateRewardValue?: number;
}

export enum KeywordType {
  CATEGORY = 'category',
  BRAND = 'brand'
}

export interface IKeyword {
  value: string;
  type: KeywordType;
}

export interface ICategory {
  name: string;
  imageUrl: string;
  lifestyleUrl: string;
  rank: number;
}

export enum MissionListType {
  DEFAULT = 'Default',
  KEEP_EARNING = 'KeepEarning',
  FEATURED = 'Featured',
  SPECIAL = 'Specials',
  JUST_FOR_YOU = 'JustForYou',
  TRENDING = 'Trending',
  DYNAMIC_LIST_2 = 'EarnList2',
  DYNAMIC_LIST_3 = 'EarnList3',
  DYNAMIC_LIST_4 = 'EarnList4',
  DYNAMIC_LIST_6 = 'EarnList6',
  EXCEPTIONAL = 'EarnList5'
}

export enum RecentSearchHistoryType {
  MISSION = 'mission',
  CATEGORY = 'category '
}
export interface IRecentSearchHistory {
  id: string;
  name: string;
  type: RecentSearchHistoryType;
  isAvailableStreakIndicator?: boolean;
}
