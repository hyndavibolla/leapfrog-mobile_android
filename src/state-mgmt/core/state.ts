import { Props as ModalProps } from '_components/Modal';
import { Props as ToastProps } from '_components/Toast';
import { IExpirePointsBannerDataSet, TooltipKey } from '_models/general';

export interface ICoreState {
  isDeepLinkListenerReady: boolean;
  isAppReady: boolean;
  isConnected: boolean;
  toastList: { key: string; props: ToastProps }[];
  modalList: { key: string; props: ModalProps }[];
  deepLink: { route: string; params: Object };
  lastRouteKey: string;
  lastRouteParams: object;
  routeHistory: string[];
  hasSeenOnboarding: boolean;
  viewedTooltipList: TooltipKey[];
  dismissedTooltipList: TooltipKey[];
  hasSeenMissionModal: boolean;
  expirePointsBannerDataSet: IExpirePointsBannerDataSet;
  roToken: string;
  marketingId: string;
  showForcedUpdateScreen: boolean;
  isTutorialVisible: boolean;
  isTutorialAvailable: boolean;
  tutorialFrom: string;
  experimentsConfig: object;
}

export const initialState: ICoreState = {
  isDeepLinkListenerReady: false,
  isAppReady: false,
  isConnected: true,
  toastList: [],
  modalList: [],
  deepLink: null,
  lastRouteKey: null,
  lastRouteParams: undefined,
  routeHistory: [],
  hasSeenOnboarding: null,
  viewedTooltipList: [],
  dismissedTooltipList: [],
  hasSeenMissionModal: false,
  expirePointsBannerDataSet: { lastAcceptedDate: 0, lastPointsDate: 0 },
  roToken: null,
  marketingId: undefined,
  showForcedUpdateScreen: false,
  isTutorialVisible: false,
  isTutorialAvailable: false,
  tutorialFrom: '',
  experimentsConfig: {}
};
