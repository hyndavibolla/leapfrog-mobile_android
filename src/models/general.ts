import { Dispatch } from 'react';
import Auth0 from 'react-native-auth0';

import { ApiService } from '_services/ApiService';
import { AppService } from '_services/AppService';
import { CCPAService } from '_services/CCPAService';
import { ICrashlyticsService } from '_services/CrashlyticsService';
import { EventTrackerService } from '_services/EventTrackerService';
import { IHttpService } from '_services/HttpService';
import { NativeHelperService } from '_services/NativeHelperService';
import { Logger } from '_services/Logger';
import { IPerformanceTrackingService } from '_services/PerformanceTrackingService';
import { RaiseService } from '_services/RaiseService';
import { RemoteConfigService } from '_services/RemoteConfigService';
import { ICardLinkState } from '_state_mgmt/cardLink/state';
import { ICoreState } from '_state_mgmt/core/state';
import { IGameState } from '_state_mgmt/game/state';
import { IGiftCardState } from '_state_mgmt/giftCard/state';
import { IMessagesState } from '_state_mgmt/messages/state';
import { IMissionState } from '_state_mgmt/mission/state';
import { IRewardState } from '_state_mgmt/reward/state';
import { ISurveyState } from '_state_mgmt/survey/state';
import { IUserState } from '_state_mgmt/user/state';

export interface IAction<T = string, P = any> {
  type: T;
  payload: P;
}

export type IReducer<S = any, A = any> = (state: S, action: IAction<A>) => S;

export interface Deps {
  apiService: ApiService;
  appService: AppService;
  crashlyticsService: ICrashlyticsService;
  httpService: IHttpService;
  performanceTrackingService: IPerformanceTrackingService;
  raiseService: RaiseService;
  nativeHelperService: NativeHelperService;
  stateSnapshot: { get: () => IGlobalState; set: (state: IGlobalState) => void };
  logger: Logger;
  eventTrackerService: EventTrackerService;
  ccpaService: CCPAService;
  remoteConfigService: RemoteConfigService;
  authService: Auth0;
}
export type IGlobalState = {
  core: ICoreState;
  user: IUserState;
  game: IGameState;
  mission: IMissionState;
  reward: IRewardState;
  survey: ISurveyState;
  messages: IMessagesState;
  cardLink: ICardLinkState;
  giftCard: IGiftCardState;
};

export interface IGlobalContext {
  state: IGlobalState;
  dispatch: Dispatch<IAction<any, any>>;
  deps: Deps;
}

export enum HTTP_METHOD {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  PATCH = 'PATCH',
  DELETE = 'DELETE'
}

export interface ITokenSet {
  accessToken: string;
  refreshToken: string;
  accessTokenExpiryTime: string;
  sywToken?: string;
  /**
   * @deprecated deviceId does not belong to ITokenSet anymore.
   */
  deviceId?: string;
  scope?: string;
  expiresIn?: number;
  tokenType?: string;
}

export enum FeatureFlag {
  API_OVERRIDE = 'API_OVERRIDE',
  TEST_ONLY = 'TEST_ONLY',
  ENV_BANNER = 'ENV_BANNER',
  BUTTON_DEBUG = 'BUTTON_DEBUG',
  LOG_LEVEL_INFO = 'LOG_LEVEL_INFO',
  LOG_LEVEL_DEBUG = 'LOG_LEVEL_DEBUG',
  LOG_LEVEL_WARN = 'LOG_LEVEL_WARN',
  LOG_LEVEL_ERROR = 'LOG_LEVEL_ERROR',
  LOG_VIEW = 'LOG_VIEW',
  STREAK_REDEEM_CARD = 'STREAK_REDEEM_CARD',
  SAILTHRU_TRACKING = 'SAILTHRU_TRACKING',
  SURVEY = 'SURVEY',
  PREDICTIVE_MAP_SEARCH = 'PREDICTIVE_MAP_SEARCH',
  MISSIONS_SEARCH = 'MISSIONS_SEARCH',
  SEE_ALL_TRANSACTIONS = 'SEE_ALL_TRANSACTIONS',
  WALLET_YOUR_GC = 'WALLET_YOUR_GC'
}

export type DateLike = Date | string | number;

export type HTMLString = string;

export interface IGameLevelUI {
  updatedAt: DateLike;
  level: number;
}

export interface ICCPASetting {
  allow: boolean;
}

export interface IOnboarding {
  seenOnboarding: boolean;
}

export enum ButtonCreativeType {
  HERO = 'hero',
  CAROUSEL = 'carousel',
  LIST = 'list',
  GRID = 'grid',
  DETAIL = 'detail',
  OTHER = 'other'
}

export interface ISearchHistoryItem {
  keyword: string;
  lastUpdatedAt: DateLike;
}

export enum TooltipKey {
  EARN = 'earn',
  REWARDS = 'rewards',
  MISSIONS = 'missions',
  STREAK = 'streak'
}

export interface IApiOverrideRule {
  key: string;
  description?: string;
  matcher: string; // reg exp
  status: number;
  response: any;
}

export namespace RemoteConfigModel {
  export interface IAppVersion {
    suggested: number;
    required: number;
  }

  export interface IEarnSection {
    name: string;
    order: number;
    visible: boolean;
    title: string;
    description: string;
    seeAllButton: boolean;
  }
}

export enum RUNTIME_ENV {
  LIVE = 'live',
  DEV_TEST = 'test'
}

export interface IMeasure {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface IExpirePointsBannerDataSet {
  lastPointsDate: DateLike;
  lastAcceptedDate: DateLike;
}

export interface ISailthruMessage {
  createdAt: string;
  custom: {
    [k: string]: string;
  };
  htmlText: string;
  id: string;
  isRead: boolean;
  text: string;
  title: string;
  type: string;
  cardImageUrl?: string;
  share: boolean;
}

export interface ITutorialStatus {
  isBannerWatched: boolean;
  date: string;
}
