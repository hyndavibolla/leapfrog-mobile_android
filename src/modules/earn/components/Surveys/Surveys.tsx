import React, { memo, useMemo, useCallback, useEffect, useContext } from 'react';
import { View, TouchableHighlight } from 'react-native';
import { useNavigation } from '@react-navigation/native';

import { Flagged } from '_components/Flagged';
import ErrorBoundary from '_components/ErrorBoundary';
import { Title, TitleType } from '_components/Title';
import { LargeSurveyCard } from '_components/LargeSurveyCard';

import { useTestingHelper } from '_utils/useTestingHelper';

import { ROUTES, TealiumEventType, PageNames, PageType, ENV, UxObject, EventDetail, ForterActionType } from '_constants';
import { GlobalContext } from '_state_mgmt/GlobalState';
import { useEventTracker } from '_state_mgmt/core/hooks';
import { useGetSurveyProfileQuestionList } from '_state_mgmt/survey/hooks';
import { KnownMissionSearchKey } from '_state_mgmt/mission/state';

import { FeatureFlag } from '_models/general';
import { Provider } from '_models/mission';
import { PromptModel } from '_models';

import { styles } from './styles';

export interface Props {
  focusKey: string;
  title?: string;
  description?: string;
  seeAllButton?: boolean;
}

const Surveys = ({ focusKey, ...props }: Props) => {
  const { getTestIdProps } = useTestingHelper('earn-main-surveys');
  const navigation = useNavigation();
  const { state } = useContext(GlobalContext);
  const { missionSearchMap, missionMap } = state.mission;
  const { trackSystemEvent } = useEventTracker();
  const [onGetSurveyQuestionList, isGettingSurveyPreQuestionList = true] = useGetSurveyProfileQuestionList();

  useEffect(() => {
    onGetSurveyQuestionList();
  }, [focusKey, onGetSurveyQuestionList]);

  const shouldShowPQButton = !isGettingSurveyPreQuestionList && !!state.survey.answerGroupMap[PromptModel.QuestionGroupName.SURVEY_PROFILE].length;

  const surveyOffer = useMemo(
    () =>
      missionSearchMap[KnownMissionSearchKey.EXCEPTIONAL].map(key => missionMap[key]).find(m => m?.provider.toLowerCase() === Provider.SURVEY.toLowerCase()),
    [missionSearchMap, missionMap]
  );

  const onSurveyPress = useCallback(() => {
    let otherTrackingAttributes = {};
    if (shouldShowPQButton || !state.survey.questionGroupMap[PromptModel.QuestionGroupName.SURVEY_PROFILE].length) {
      otherTrackingAttributes = { event_type: PageType.SURVEY_DETAIL, exit_link: surveyOffer?.callToActionUrl };
      navigation.navigate(ROUTES.SURVEY_DETAIL, { surveyUri: surveyOffer?.callToActionUrl });
    } else {
      otherTrackingAttributes = { event_type: PageType.SURVEY_PQ };
      navigation.navigate(ROUTES.SURVEY_PQ);
    }

    trackSystemEvent(
      TealiumEventType.SURVEY_CARD_TAP,
      {
        page_name: PageNames.MAIN.EARN,
        page_type: PageType.INFO,
        address: `${ENV.SCHEME}${ROUTES.SURVEY_DETAIL}`,
        uxObject: UxObject.TILE,
        event_detail: EventDetail.OPEN,
        ...otherTrackingAttributes
      },
      ForterActionType.TAP
    );
  }, [surveyOffer, trackSystemEvent, shouldShowPQButton]); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <View style={styles.sectionMain} {...getTestIdProps('container')}>
      <Flagged feature={FeatureFlag.SURVEY}>
        {!surveyOffer ? null : (
          <ErrorBoundary>
            <View style={styles.surveySectionHeader} {...getTestIdProps('survey-header')}>
              <Title type={TitleType.SECTION} numberOfLines={1} ellipsizeMode="tail">
                {props?.title ?? 'Surveys'}
              </Title>
              {!shouldShowPQButton ? null : (
                <TouchableHighlight onPress={() => navigation.navigate(ROUTES.SURVEY_PQ)} underlayColor="transparent" {...getTestIdProps('pq-btn')}>
                  <Title type={TitleType.SECTION} style={styles.link}>
                    Edit preferences
                  </Title>
                </TouchableHighlight>
              )}
            </View>
            <View>
              <LargeSurveyCard survey={surveyOffer} onPress={onSurveyPress} />
            </View>
          </ErrorBoundary>
        )}
      </Flagged>
    </View>
  );
};

export default memo(Surveys);
