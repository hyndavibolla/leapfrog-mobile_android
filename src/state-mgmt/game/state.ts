import { GameModel } from '../../models';

export interface IGameState {
  current: GameModel.IGame;
  historical: { totalLifetimeEarnedPoints: number; pointByMonthMap: { [month: string]: number } };
}

export const initialState: IGameState = {
  current: {
    balance: {
      availablePoints: 0,
      pendingPoints: 0,
      memberOwnPointsToExpire: []
    },
    memberships: {
      userHasSywCard: false,
      userIsSywMaxMember: false,
      userIsAllowedToUseSywMax: true,
      sywCardLastFour: '1234'
    },
    missions: {
      pointsPerCent: 0
    },
    lastUpdatedAt: 0
  },
  historical: {
    totalLifetimeEarnedPoints: 0,
    pointByMonthMap: {}
  }
};
