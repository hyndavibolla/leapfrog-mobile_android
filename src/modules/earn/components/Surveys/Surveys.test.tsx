import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react-native';

import Surveys, { Props } from './Surveys';

import { getInitialState } from '_state_mgmt/GlobalState';
import { Provider } from '_models/mission';
import { Deps, IGlobalState } from '_models/general';
import { PromptModel } from '_models';
import { KnownMissionSearchKey } from '_state_mgmt/mission/state';
import { getAnswer_1, getMission_1, getQuestion_1, getQuestion_2 } from '_test_utils/entities';
import { getMockDeps } from '_test_utils/getMockDeps';
import { renderWithGlobalContext } from '_test_utils/renderWithGlobalContext';
import { createUUID } from '_utils/create-uuid';

import { ROUTES } from '_constants';

const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  __esModule: true,
  useNavigation: () => ({ navigate: mockNavigate })
}));

describe('Surveys', () => {
  let props: Props;
  let deps: Deps;
  let initialState: IGlobalState;

  beforeEach(() => {
    deps = getMockDeps();
    initialState = getInitialState();

    initialState.mission.missionSearchMap[KnownMissionSearchKey.EXCEPTIONAL] = ['test12345'];
    initialState.mission.missionMap = { ['test12345']: { ...getMission_1(), provider: Provider.SURVEY } };
    initialState.survey.questionGroupMap = { [PromptModel.QuestionGroupName.SURVEY_PROFILE]: [getQuestion_1(), getQuestion_2()] };
    initialState.survey.answerGroupMap = { [PromptModel.QuestionGroupName.SURVEY_PROFILE]: [] };
    props = {
      focusKey: createUUID()
    };

    deps.apiService.fetchMissionList = jest.fn().mockImplementationOnce(({ listType }) => {
      return {
        userId: 'userId',
        missions: [
          {
            ...getMission_1(),
            provider: Provider.SURVEY
          }
        ],
        listType
      };
    });
  });

  it('should render', async () => {
    const { toJSON } = renderWithGlobalContext(<Surveys {...props} />);
    await waitFor(() => expect(toJSON()).toMatchSnapshot());
  });

  it('should navigate to survey pq when a survey card was pressed', async () => {
    deps.apiService.fetchPromptList = jest.fn().mockImplementationOnce(() => {
      return { answeredQuestions: [], promptQuestions: [getQuestion_1()] };
    });

    const { findByTestId } = renderWithGlobalContext(<Surveys {...props} />, deps, initialState);
    fireEvent.press(await findByTestId('large-survey-card-container'));
    expect(mockNavigate).toBeCalledWith(ROUTES.SURVEY_PQ);
  });

  it('should have a pressable survey card when PQ is NOT needed', async () => {
    deps.apiService.fetchPromptList = jest.fn().mockImplementationOnce(() => {
      return { answeredQuestions: [getAnswer_1()], promptQuestions: [getQuestion_1()] };
    });

    const { findByTestId } = renderWithGlobalContext(<Surveys {...props} />, deps, initialState);
    fireEvent.press(await findByTestId('large-survey-card-container'));
    expect(mockNavigate).toBeCalledWith(ROUTES.SURVEY_DETAIL, {
      surveyUri: getMission_1().callToActionUrl
    });
  });

  it('should have an edit pq questions when there are any answered question to edit', async () => {
    deps.apiService.fetchPromptList = jest.fn().mockImplementationOnce(() => {
      return { answeredQuestions: [getAnswer_1()], promptQuestions: [getQuestion_1()] };
    });

    const { findByTestId } = renderWithGlobalContext(<Surveys {...props} />, deps, initialState);
    fireEvent.press(await findByTestId('earn-main-surveys-pq-btn'));
    expect(mockNavigate).toBeCalledWith(ROUTES.SURVEY_PQ);
  });

  it('should NOT have a edit pq questions when there is not any answered question to edit', async () => {
    deps.apiService.fetchPromptList = jest.fn().mockImplementationOnce(() => {
      return { answeredQuestions: [], promptQuestions: [getQuestion_1()] };
    });

    const { queryByTestId } = renderWithGlobalContext(<Surveys {...props} />, deps, initialState);

    await waitFor(() => {
      expect(queryByTestId('earn-main-surveys-pq-btn')).toBeNull();
    });
  });
});
