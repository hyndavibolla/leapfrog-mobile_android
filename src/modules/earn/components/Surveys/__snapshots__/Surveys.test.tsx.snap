// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Surveys should render 1`] = `
<View
  accessibilityLabel="earn-main-surveys-container"
  accessible={true}
  style={
    Object {
      "paddingBottom": 24,
      "paddingHorizontal": 16,
    }
  }
  testID="earn-main-surveys-container"
/>
`;
